src/main.rs

```diff
@@ -204,6 +214,43 @@ async fn main() -> Result<(), Box<dyn Error>> {
         shutdown_tx.send(()).await.ok();
     });
 
+    let mempool_shard_stores = node.shard_stores.clone();
+    let mempool_shard_senders = node.shard_senders.clone();
+
+    let mempool_block_store = block_store.clone();
+
+    tokio::spawn(async move {
+        let l1_client: Option<Box<dyn L1Client>> = match RealL1Client::new(app_config.l1_rpc_url) {
+            Ok(client) => Some(Box::new(client)),
+            Err(_) => None,
+        };
+        let service = MyHubService::new(
+            mempool_block_store,
+            mempool_shard_stores,
+            mempool_shard_senders,
+            statsd_client.clone(),
+            app_config.consensus.num_shards,
+            Box::new(routing::ShardRouter {}),
+            l1_client,
+        );
+
+        loop {
+            if let Some(mempool_message) = mempool_rx.recv().await {
+                match mempool_message {
+                    MempoolMessage::UserMessage(message) => {
+                        let result = service.submit_message_internal(message, false).await;
```

Comment:
This seemed like the most sane way to share the validation logic between RPC-submitted messages and gossip-submitted messages

Reply:
The dependency is inverted. The mempool should live one level below the service, so it shouldn't have any knowledge of the service. The service should be submitting to the mempool. See the new structure in https://github.com/farcasterxyz/snapchain/pull/203